src/ripple/app/tx/impl/Transactor.cpp

```diff
@@ -775,6 +922,541 @@ Transactor::reset(XRPAmount fee)
+            // check if the TSH exists and/or has any hooks
+            auto tshAcc = view.peek(keylet::account(tshAccountID));
+            if (!tshAcc)
+                continue;
+
+            // compute and deduct fees for the TSH if applicable
+            FeeUnit64 tshFee =
+                calculateHookChainFee(view, ctx_.tx, klTshHook, !canRollback);
+
+            // no hooks to execute, skip tsh
+            if (tshFee == 0)
+                continue;
+                
+            XRPAmount tshFeeDrops = view.fees().toDrops(tshFee);
+            assert(tshFeeDrops >= beast::zero);
+
+            STAmount priorBalance = tshAcc->getFieldAmount(sfBalance);
+
+            if (canRollback)
+            {
+                // this is not a collect call so we will force the tsh's fee to 0
+                // the otxn paid the fee for this tsh chain execution already.
+                tshFeeDrops = 0;
+            }
+            else
+            {
+                // this is a collect call so first check if the tsh can accept
+                uint32_t tshFlags = tshAcc->getFieldU32(sfFlags);
+                if (!canRollback && !(tshFlags & lsfTshCollect))
+                {
+                    // this TSH doesn't allow collect calls, skip
+                    JLOG(j_.trace())
+                        << "HookInfo[" << account_ << "]: TSH acc " << tshAccountID << " "
+                        << "hook chain execution skipped due to lack of lsfTshCollect flag.";
+                    continue;
+                }
+
+                // now check if they can afford this collect call
+                auto const uOwnerCount = tshAcc->getFieldU32(sfOwnerCount);
+                auto const reserve = view.fees().accountReserve(uOwnerCount);
+
+                if (tshFeeDrops + reserve > priorBalance)
+                {
+                    JLOG(j_.trace())
+                        << "HookInfo[" << account_ << "]: TSH acc " << tshAccountID << " "
+                        << "hook chain execution skipped due to lack of TSH acc funds.";
+                    continue;
+                }
+            }
+
+            if (tshFeeDrops > beast::zero)
+            {
+                STAmount finalBalance = priorBalance - tshFeeDrops;
+                assert(finalBalance >= beast::zero);
+                assert(finalBalance < priorBalance);
+
+                tshAcc->setFieldAmount(sfBalance, finalBalance);
+                view.update(tshAcc);
+                ctx_.destroyXRP(tshFeeDrops);
+            }
+        }
+
+        // execution to here means we can run the TSH's hook chain
+        TER tshResult =
+            executeHookChain(
+                tshHook,
+                stateMap,
+                results,
+                tshAccountID,
+                strong,
+                provisionalMeta);
+
+        if (canRollback && (tshResult != tesSUCCESS))
+            return tshResult;
+    }
+
+    return tesSUCCESS;
+}
+
+void
+Transactor::doAaw(
```

Comment:
Can we have a slightly more descriptive name here?

Reply:
Yep changing it back to `doAgainAsWeak`. Not sure why I thought that was a bad name to begin with 😅